### ISSUE TITLE
Extra lines of context above/below search results

### ISSUE THREAD

**Issue by andrewchilds** [atom/find-and-replace#190]
ST includes two lines above and below search results for context, which I regularly found useful and would love to see included in Atom.

Ideally the line count would be a setting instead of hardcoded:

```
[ ] Focus Editor After Search
[ ] Open Project Find Results In Right Pane
[2] Lines of Context Above/Below Search Results
```

:beers:

**Comment by mbillard** [comment]
:+1:

**Comment by zenlambda** [comment]
+1

**Comment by MWeeden43** [comment]
:+1:

**Comment by gabmontes** [comment]
Would be awesome to have this enhancement in.

I think most of us can do it and send a PR, anyway. Any clue on where to start?

**Comment by jayasurian123** [comment]
This is really a missing item. :+1:

**Comment by wcurtis** [comment]
:+1: As a fresh convert from Sublime Text I'd love to see this added to Atom. The extra few lines of context really make a difference. Here's a comparison for reference.

Atom 1.0
![screen shot 2015-07-10 at 4 39 40 pm](https://cloud.githubusercontent.com/assets/530242/8628588/5a7fc6d4-2722-11e5-8f08-d1002a1d7de6.png)

Sublime Text 3
![screen shot 2015-07-10 at 4 44 19 pm](https://cloud.githubusercontent.com/assets/530242/8628667/f310d262-2722-11e5-9010-281d5f1f4d38.png)

**Comment by cvogt** [comment]
+1

**Comment by zombiecrab** [comment]
:thumbsup: Really need this.

**Comment by lukeaus** [comment]
+1

**Comment by barteko** [comment]
+1

**Comment by ReneMoraales** [comment]
+1 PARFAVAR

**Comment by kaushikgopal** [comment]
+1

**Comment by inelsonrocha** [comment]
:+1:  (as this is the only real argument preventing global atom adoption in my company... instead of ST)

**Comment by DouweM** [comment]
:+1:

**Comment by 807811b0-1373-11e5-b60b-1697f925ec7b** [comment]
If anyone has a fork of Find and Replace that adds this functionality, please comment.

**Comment by rosepuppy** [comment]
:+1:

**Comment by rosepuppy** [comment]
:+1:

**Comment by seedofjoy** [comment]
:+1:

**Comment by alxpy** [comment]
:+1:

**Comment by ddenis-ua** [comment]
+1

**Comment by antonpaisov** [comment]
:+1:

**Comment by jchlu** [comment]
+1

**Comment by borb** [comment]
+1

This would be *incredibly* useful - right now, I drop to a shell and use grep -A3 -B3.

**Comment by maprihoda** [comment]
+1

**Comment by dmo-odoo** [comment]
Hello,

This will nearly be 2 years since this issue has been opened, and it is still the most bothering everyday problem with atom in my opinion.

Could you please at least aknowledge this issue, even the answer is "we think it's useless, we won't do it" ?

Thank you,
David

**Comment by nim-odoo** [comment]
That's definitively THE missing feature of atom...

**Comment by kaushikgopal** [comment]
@dmo-odoo this has been tagged as an "enhancement" which basically means don't hold your breath for this to get "fixed". Unless it's a bug there can't be an expectation for it to be fixed.


Below is complete speculation on my part: 

There could be performance considerations which is why this feature is not built in. I mean, it's clear that the folks at Atom have noticed this feature in Sublime Text. If it's not already there, there's a reason why.

Even better: it's totally open source! you can see the parts of the code that enable this feature. Add the enhancement yourself and send in a PR. That should kick start the process again. If you're not sure how, you'll have to be patient for someone else to do it. That's unfortunately how the game works.

**Comment by 50Wliu** [comment]
> Could you please at least aknowledge this issue, even the answer is "we think it's useless and we won't do it" ?

@dmo-odoo The issue is still open :wink:, it's just that no one from the Atom team (or the community!) has decided to work on this request yet.

@kaushikgopal is correct in that sending a PR is probably the best way to revive interest in this issue.  I'll be tagging this issue as `help-wanted`.

**Comment by MartinMuzatko** [comment]
https://github.com/atom/find-and-replace/blob/master/lib%2Fproject%2Fmatch-view.coffee#L11-L12

This is where you can change the display of the lines. I'll give it a shot but I'm not familiar with the build system of atom packages and I found no documentation so far.

**Comment by MartinMuzatko** [comment]
I got the extra lines, now I only need to properly format it

```js
    lines = atom.project.findBufferForPath(filePath)?.lines

    prefixLines = lines.slice(range.start.row - 2, range.start.row)
    suffixLines = lines.slice(range.end.row, range.end.row + 2)
    # if range.start.row <= 2
    #   prefixLines = lines.slice(0, range.start.row)
    # if range.end.row >= lines.length - 2
    #   suffixLines = lines.slice(range.end.row, lines.length)

    @li class: 'search-result list-item', =>
      @span range.start.row + 1, class: 'line-number text-subtle'
      @span => @raw prefixLines.join('<br>')
      @span class: 'preview', outlet: 'preview', =>
        @span prefix
        @span match.matchText, class: 'match highlight-info', outlet: 'matchText'
        @span match.matchText, class: 'replacement highlight-success', outlet: 'replacementText'
        @span suffix
      @span => @raw suffixLines.join('<br>')

```

**Comment by MartinMuzatko** [comment]
Like this :D

![f r2](https://cloud.githubusercontent.com/assets/2950505/14383049/b2b1336e-fd92-11e5-8a23-2aed4047717e.png)

**Comment by DouweM** [comment]
@MartinMuzatko You make me very happy

**Comment by cvogt** [comment]
@MartinMuzatko You are awesome!

**Comment by MartinMuzatko** [comment]
:blush: thanks guys, I do my best to fix this.

**Comment by fiz** [comment]
:+1:

**Comment by rga-odoo** [comment]
:+1:  it's really missing feature

**Comment by alexhudici** [comment]
+1

**Comment by alexhudici** [comment]
👍

**Comment by isuraed** [comment]
That looks great! Hope it's in the product soon.

**Comment by MartinMuzatko** [comment]
@isuraed it is not that easy, there are a lot of things to take care of, but it is possible.

**Comment by bmherold** [comment]
👍 💝

**Comment by rahlzel** [comment]
:+1:

**Comment by thebarty** [comment]
+1 PLEASSEEEEEEEE 👍

**Comment by MartinMuzatko** [comment]
I think I'll try it again during my winter vacation. As said: there are a few more aspects to it. but maybe finally I can finish it up.

**Comment by thebarty** [comment]
@MartinMuzatko: oh man that would be sooooo nice!!!! 👍 I am really missing this after switching over from sublime. Sublime does such a great job at this. It is just so much easier to find the right places you are searching for.

**Comment by thebarty** [comment]
Just thinking: a good reference on how to layout the results is the way that github displays search restults, p.e. here https://github.com/meteor/todos/search?utf8=%E2%9C%93&q=.call

**Comment by spencerhakim** [comment]
Not sure if it's truly related to this issue, but I hate seeing the same line printed twice (or more) when the word I'm searching for just happens to appear twice (or more) in the same line. Example: 
<img width="221" alt="capturfiles-201612341_111240" src="https://cloud.githubusercontent.com/assets/606857/20934773/5a6048cc-bbaa-11e6-955b-4ed4cb38e116.png">
I really want to see the surrounding lines in my search results, but it would be _really_ annoying if instead of seeing the same **one** line printed twice, I see the same **five** lines printed twice.

Just something to keep in mind for whoever ends up working on this (and a pre-emptive Thanks to @MartinMuzatko if you do end up working on this over the holidays)

**Comment by 50Wliu** [comment]
@spencerhakim That was already discussed and I believe the consensus was that code-wise, it would be the most logical to tackle it separately from this one.

**Comment by spencerhakim** [comment]
@50Wliu Ah, that's what I get for not reading through all the comments. Thanks for getting me up to date.

**Comment by dirk-thomas** [comment]
I thought about the way PR #702 is implementing the feature and am a bit worried that the overhead due to opening a buffer for every result might be high in cases of large scale (many search result / search results in large files). Therefore I described an alternative approach In #827. Any feedback on this would be highly welcome - especially any experience on the overhead of creating buffers for each result would be valuable to make an informed decision.

**Comment by PixelT** [comment]
👍 👍 👍

**Comment by dirk-thomas** [comment]
This has been implemented in #847 which has been merged for the upcoming beta of Atom 1.17. So I think this can be closed.
